### DISCUSSION TITLE
Как исполняется асинхронный/синхронный код внутри Libuv?

### DISCUSSION THREAD

**Discussion by ik1llo** [HowProgrammingWorks/LiveQA#33] [Q&A] [answered]
Здравствуйте, помогите пожалуйста, а то разные источники дают разную инфу, а в доках не нашел. Заранее спасибо!

Libuv когда увидит что операция асинхронная кинет ее на выполнение epoll/iocp и тд (без попадания в event loop). После того как операция выполнена прийдёт событие и коллбек закинется в очередь фазы, как я понимаю poll, как исполнится этот код? У libuv есть просто доступ к байт-коду или машинному коду, который ранее был создан, и загружен в память, просто его на этом этапе вызывает из памяти с возвращенным значением, правильно? И такой же вопрос с синхронной операцией что происходит в этот момент?

Синхронный пример:

```
console.log(1)
fs.someSyncMethod(() => console.log(2))
console.log(3) 
```

Вывод будет таким: 1,2,3.

Вопрос и по этому, что происходит в Libuv когда в него попадет синхронный код? И каким образом он может заблокировать event loop если в него попадают только выполненные коллбэки? Если я что то неправильно говорю, прошу поправить. И третий вопрос - правильно ли я понимаю что код типа простых циклов/условий/console.log() исполняется внутри V8, а синхронные/асинхронные методы биндингов (того же fs) в Libuv. Если да то как Libuv запускает их?

**Comment by tshemsedinov** [comment, marked as answer]
Код исполняется в `V8`, а не в `libuv`, вообще, как работает синхронный и асинхронный код можно понять без `libuv`, это дополнительные знания, которые конечно помогают в понимании всей картины, но совершенно не нужны для написания правильного, качественного и надежного кода. Синхронный код исполняется последовательно, в приведенном примере сначала первый `console.log`, потом вызов функции, из нее вызыввется колбек и из него срабатывает второй `console.log` потом return из функции и потом третий `console.log`. Синхронный код называют "блокирующим", потому, что во время его исполнения не могут приходить колбеки от асинхронных операций, ресолвиться промисы, таймеры не работают, они тоже на колбеках, просто частный случай. Чтоб этот код был асинхронным и не блокирующим, то вместо функции `someSyncMethod` нужно вызвать асинхронную. Ее работа немного иначе устроена, асинхронные функции сохраняют себе ссылку на колбек и возвращают управление наружу. То есть, `console.log` выведет не `1,2,3`, а сначала `1,2`, а потом, когда функция дернет колбек, после определенного ожидания, то выведется `3`. Вот тут мы дошли до `event loop`, и он может быть разным, не обязательно `libuv`, как в нодею В браузере `event loop` реализован другой библиотекой `libevent`, но суть у них одинаковая, это дернуть колбек, когда для него будут готовы данные, приходящие из внешних источников и операционной системы (т.е. из I/O).

P.S. Весь код выполняется в `V8`, ничего не передается в `libuv`, `libevent`, они просто инициируют вызовы, т.е. подсказывают `V8`, что нужно вызвать функцию (поместить ее на стек вызовов) в определенное время.

- Чтобы подробнее узнать про `event loop` смотри: https://youtu.be/8aGhZQkoFbQ
- Еще одно полезное видео https://youtu.be/cCOL7MC4Pl0
- Полый курс по асинхронному программированию тут: https://github.com/HowProgrammingWorks/Index/blob/master/Courses/Asynchronous.md
- Курс по ноде: https://github.com/HowProgrammingWorks/Index/blob/master/Courses/NodeJS.md

Еще ответ на этот вопрос: https://github.com/HowProgrammingWorks/LiveQA/discussions/15

**Reply by demimurych** [reply]
Event Loop не имеет никакого отношения к JS.  Никакого Even Loop в JS тет.
Event Loop это часть стандарта HTML5.
HTML5 выступает HOST системой в браузере для выполнения JS. 
По этой причине Event Loop, оказывает влияние на работу JS в браузере. 
Node это распространенная HOST система для JS вне браузера. 
В Node есть свой Event Loop который пытается повторять стандарт HTML5. Потому Event Loop в Node и в браузере обеспечивающем поддержку HTML5 Event Loop схожи.

В самом JS никакого Event Loop нет. 
HOST системой может выступать что угодно. Например V8 - движек выполенения JS от Google может быть HOST системой для самого себя. Соотвественно, когда V8 выступает HOST системой - никакого Event Loop нет.

Термины: _блокирующий код_ и _неблокирующий код_ никакого отношения к JS не имеет, по причине того,
что это терминология специфической HOST системы реализующей стандарт HTML5.

В JS есть REALM, HOST и JOBS. 
В одну единицу времени может выполняться только один JOBS в рамках одного HOST. 
JOBS это логическая единица, оформленная соотвествущим спецификации образом. 
Например доступ к свойству обьекта это один JOBS.
Как следствие, в JS выполнение кода может быть прервано после выполнения любой команды выступающей как JOBS. **Дефакто после завершения любой команды, поток исполнения может быть передан куда то еще.** 

Это поведение может быть изменено HOST системой, которая имеет право определять JOBS самостоятельно. 
По этой причине, в HTML5 так называемый основной тред блокирует выполнение прочего кода - HOST система весь этот код 
определяет как один JOBS. Продиктовано это несовершенством современного стандарта рендера HTML кода.

**Итого**
Всегда нужно помнить что поведение JS напрямую зависит от того кто выступает HOST для его исполнения.
Нужно четко отдавать себе отчет в том, что именно является частью стандарта JS, 
а что является специфической ососбенностью реализации конкретной HOST системы.


Настоятельно рекомендую очень внимательно [читать официальную спецификацию](https://tc39.es/ecma262/#sec-jobs)
